File: 06-linked_list/00002-Add_Two_Numbers/00002-add_two_numbers.go
// 2: Add Two Numbers
// https://leetcode.com/problems/add-two-numbers/

package main

import "fmt"

type ListNode struct {
    val int
    next *ListNode
}

type LinkedList struct {
    head *ListNode
}

func llInput(head **ListNode, inputs []int) {
    for i:=len(inputs)-1; i>=0; i-- {
        node := &ListNode{inputs[i], *head}
        *head = node
    }
}

func llOutput(head *ListNode) {
    for head != nil {
        fmt.Print(head.val, " -> ")
        head = head.next
    }
    fmt.Println("NULL")
}

// SOLUTION
func addTwoNumbers(l1, l2 *ListNode) *ListNode {
    head := &ListNode{}
	t := head
	c := 0

	for c!=0 || l1!=nil || l2!=nil {
		if l1!=nil {
			c += l1.val
			l1 = l1.next
		}
		if l2!=nil {
			c += l2.val
			l2 = l2.next
		}

		t.next = &ListNode{val: c%10}
		t = t.next
		c /= 10
	}

	return head.next
}

func main() {
    ll1, ll2 := LinkedList{}, LinkedList{}
    
    // INPUT
    li1, li2 := []int{2,4,3}, []int{5,6,4}
    llInput(&ll1.head, li1)
    llInput(&ll2.head, li2)

    // OUTPUT
    result := addTwoNumbers(ll1.head, ll2.head)
    llOutput(result)
}

File: 06-linked_list/00002-Add_Two_Numbers/00002-add_two_numbers.java
// 2: Add Two Numbers
// https://leetcode.com/problems/add-two-numbers/


class Solution {
    public static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public static class LinkedList {
        ListNode head;
        LinkedList() { this.head = null; }
    }

    public static void llInput(LinkedList list, int[] inputs) {
        for (int i=inputs.length-1; i>=0; i--) {
            ListNode node = new ListNode(inputs[i], list.head);
            list.head = node;
        }
    }

    public static void llOutput(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " -> ");
            head = head.next;
        }
        System.out.println("NULL");
    }


    // SOLUTION
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode();
        ListNode t = head;
        int c = 0;

        while (c!=0 || l1!=null || l2!=null) {
            c += (l1!=null ? l1.val : 0) + (l2!=null ? l2.val : 0);

            ListNode temp = new ListNode();
            temp.val = c%10;
            temp.next = null;

            t.next = temp;
            t = t.next;
            c /= 10;

            if (l1!=null) l1 = l1.next;
            if (l2!=null) l2 = l2.next;
        }

        return head.next;
    }


    public static void main(String[] args) {
        Solution o = new Solution();
        LinkedList ll1 = new LinkedList();
        LinkedList ll2 = new LinkedList();

        // INPUT
        int[] li1 = {2,4,3};
        int[] li2 = {5,6,4};
        llInput(ll1, li1);
        llInput(ll2, li2);

        // OUTPUT
        var result = o.addTwoNumbers(ll1.head, ll2.head);
        llOutput(result);
    }
}